tests/pytests/unit/spm/test_spm.py

```diff
@@ -0,0 +1,211 @@
+                "description: {description}"
+            ).format(**_F1["definition"]),
+        ),
+        ("modules/mod1.py", "# mod1.py"),
+        ("modules/mod2.py", "# mod2.py"),
+        ("states/state1.sls", "# state1.sls"),
+        ("states/state2.sls", "# state2.sls"),
+    )
+    return _F1
+
+
+class SPMTestUserInterface(salt.spm.SPMUserInterface):
+    """
+    Unit test user interface to SPMClient
+    """
+
+    def __init__(self):
+        self._status = []
+        self._confirm = []
+        self._error = []
+
+    def status(self, msg):
+        self._status.append(msg)
+
+    def confirm(self, action):
+        self._confirm.append(action)
+
+    def error(self, msg):
+        self._error.append(msg)
+
+
+@pytest.fixture()
+def setup_spm(tmp_path):
+    minion_config = salt.config.DEFAULT_MINION_OPTS.copy()
+    minion_config.update(
+        {
+            "spm_logfile": str(tmp_path / "log"),
+            "spm_repos_config": str(tmp_path / "etc" / "spm.repos"),
+            "spm_cache_dir": str(tmp_path / "cache"),
+            "spm_build_dir": str(tmp_path / "build"),
+            "spm_build_exclude": [".git"],
+            "spm_db_provider": "sqlite3",
+            "spm_files_provider": "local",
+            "spm_db": str(tmp_path / "packages.db"),
+            "extension_modules": str(tmp_path / "modules"),
+            "file_roots": {"base": [str(tmp_path)]},
+            "formula_path": str(tmp_path / "spm"),
+            "pillar_path": str(tmp_path / "pillar"),
+            "reactor_path": str(tmp_path / "reactor"),
+            "assume_yes": True,
+            "root_dir": str(tmp_path),
+            "force": False,
+            "verbose": False,
+            "cache": "localfs",
+            "cachedir": str(tmp_path / "cache"),
+            "spm_repo_dups": "ignore",
+            "spm_share_dir": str(tmp_path / "share"),
+        }
+    )
+    ui = SPMTestUserInterface()
+    client = salt.spm.SPMClient(ui, minion_config)
+    minion_opts = salt.config.DEFAULT_MINION_OPTS.copy()
+    return tmp_path, ui, client, minion_config, minion_opts
+
+
+def _create_formula_files(formula, _tmp_spm):
+    fdir = str(_tmp_spm / formula["definition"]["name"])
+    shutil.rmtree(fdir, ignore_errors=True)
+    os.mkdir(fdir)
+    for path, contents in formula["contents"]:
+        path = os.path.join(fdir, path)
+        dirname, _ = os.path.split(path)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+        with salt.utils.files.fopen(path, "w") as f:
+            f.write(contents)
+    return fdir
+
+
+def test_build_install(setup_spm, f1_content):
+    # Build package
+    _tmp_spm, ui, client, minion_config, minion_opts = setup_spm
+    fdir = _create_formula_files(f1_content, _tmp_spm)
+    with patch("salt.client.Caller", MagicMock(return_value=minion_opts)):
+        with patch(
+            "salt.client.get_local_client",
+            MagicMock(return_value=minion_opts["conf_file"]),
+        ):
```

Comment:
**suggestion (non-blocking)** to avoid a double-nested with, it's possible to define the patch and then use it in a with block:

```
patch_caller = patch("salt.client.Caller", return_value=minion_opts, autospec=True)
patch_get_client = patch("salt.client.get_local_client", return_value=minion_opts['conf_file'], autospec=True)
with patch_caller, patch_get_client:
    client.run(['build', fdir])
```

For instance.